sycl/source/detail/usm/usm_impl.cpp

```diff
@@ -226,5 +226,70 @@ void *aligned_alloc(size_t Alignment, size_t Size, const queue &Q, alloc Kind) {
   return aligned_alloc(Alignment, Size, Q.get_device(), Q.get_context(), Kind);
 }
 
+// Pointer queries
+/// Query the allocation type from a USM pointer
+/// Returns alloc::host for all pointers in a host context.
+///
+/// @param ptr is the USM pointer to query
+/// @param ctxt is the sycl context the ptr was allocated in
+alloc get_pointer_type(const void *Ptr, const context &Ctxt) {
```

Comment:
What if user mixes up context used for USM allocation and context passed to get_pointer_type function?
```
sycl::context Ctx1(sycl::cpu_selector), Ctx2(sycl::host_selector);
auto *Ptr = sycl::alloc_shared(Ctx1, ...);
auto Type = sycl::get_pointer_type(Ptr, Ctx2); // <- This returns alloc::host.
```
Can we add some diagnostic to catch such issues?

Reply:
The PI_CALL should fail if something like that goes wrong.

Reply:
In the example above get_pointer_type doesn't call PI.

Reply:
I'm not sure I follow.

Reply:
I mean, the code snippet doesn't return any error, while being illegal because Ptr is allocated in the Ctx1 but get_pointer_type is called with Ctx2.

Reply:
I'm not sure this can really be handled. In this case, I would expect it to just work as shared allocations work on the host and the host doesn't really care what context anything is in. CPU memory is CPU memory.

The trickier case is:
```
sycl::context Ctx1(sycl::gpu_selector), Ctx2(sycl::host_selector);
auto *Ptr = sycl::alloc_device(Ctx1, ...);
auto Type = sycl::get_pointer_type(Ptr, Ctx2); // <- This returns alloc::host.
```
Trying to use Ptr on the host ought to crash the program or give garbage values. We would have to keep track of insane amounts of state to try to get a host device context to bail on this.

Reply:
I think we should invent some diagnostic, probably enabled if some option is set only.
People facing this issue in real code already.

Reply:
I'm OK if done as a separate patch.